pandas/tests/series/test_analytics.py

```diff
@@ -225,101 +225,143 @@ def test_median(self):
         int_ts = Series(np.ones(10, dtype=int), index=lrange(10))
         tm.assert_almost_equal(np.median(int_ts), int_ts.median())
 
-    def test_mode(self):
-        # No mode should be found.
-        exp = Series([], dtype=np.float64)
-        tm.assert_series_equal(Series([]).mode(), exp)
-
-        exp = Series([1], dtype=np.int64)
-        tm.assert_series_equal(Series([1]).mode(), exp)
-
-        exp = Series(['a', 'b', 'c'], dtype=np.object)
-        tm.assert_series_equal(Series(['a', 'b', 'c']).mode(), exp)
+    @pytest.mark.parametrize('dropna, expected', [
+        (True, Series([], dtype=np.float64)),
+        (False, Series([], dtype=np.float64))
+    ])
+    def test_mode_empty(self, dropna, expected):
+        s = Series([], dtype=np.float64)
+        tm.assert_series_equal(s.mode(dropna), expected)
 
-        # Test numerical data types.
-        exp_single = [1]
+    @pytest.mark.parametrize('dropna, expected1, expected2, expected3', [
+        (True, [1], [1, 3], [1.0]),
+        (False, [1], [1, 3], [1, np.nan]),
+    ])
+    def test_mode_numerical(self, dropna, expected1, expected2, expected3):
         data_single = [1] * 5 + [2] * 3
-
-        exp_multi = [1, 3]
         data_multi = [1] * 5 + [2] * 3 + [3] * 5
 
         for dt in np.typecodes['AllInteger'] + np.typecodes['Float']:
             s = Series(data_single, dtype=dt)
-            exp = Series(exp_single, dtype=dt)
-            tm.assert_series_equal(s.mode(), exp)
+            expected1 = Series(expected1, dtype=dt)
+            tm.assert_series_equal(s.mode(dropna), expected1)
 
             s = Series(data_multi, dtype=dt)
-            exp = Series(exp_multi, dtype=dt)
-            tm.assert_series_equal(s.mode(), exp)
+            expected2 = Series(expected2, dtype=dt)
+            tm.assert_series_equal(s.mode(dropna), expected2)
 
+        s = Series([1, np.nan])
+        expected = Series(expected3)
+        tm.assert_series_equal(s.mode(dropna), expected)
+
+    @pytest.mark.parametrize('dropna, expected1, expected2, expected3', [
+        (True, ['b'], ['bar'], ['nan']),
+        (False, ['b'], ['bar', np.nan], ['nan'])
```

Comment:
why is the expected string nan here? that's really odd

Reply:
I think it's because when ``dtype=str`` is specified then ``np.nan`` gets converted to ``'nan'`` and treated like any other string. In this case the ``dropna`` parameter has no effect. The same thing happens with ``value_counts()``:
```
In [2]: data = ['foo', 'bar', 'bar', np.nan, np.nan, np.nan]
In [3]: s = pd.Series(data, dtype=str)

In [4]: s.mode(dropna=True)
Out[4]:
0    nan                         
dtype: object

In [5]: s.mode(dropna=False)
Out[5]:
0    nan                 
dtype: object  

In [6]: s.value_counts(dropna=True)
Out[6]:
nan    3 
bar    2 
foo    1                                        
dtype: int64

In [7]: s.value_counts(dropna=False)
Out[7]:
nan    3 
bar    2 
foo    1                                        
dtype: int64
```